src/Stratis.Features.FederatedPeg/TargetChain/WithdrawalTransactionBuilder.cs

```diff
@@ -25,28 +38,42 @@ public class WithdrawalTransactionBuilder : IWithdrawalTransactionBuilder
         private readonly IFederationWalletManager federationWalletManager;
         private readonly IFederationWalletTransactionHandler federationWalletTransactionHandler;
         private readonly IFederatedPegSettings federatedPegSettings;
+        private readonly IStateRepositoryRoot stateRepositoryRoot;
 
         public WithdrawalTransactionBuilder(
             ILoggerFactory loggerFactory,
             Network network,
             IFederationWalletManager federationWalletManager,
             IFederationWalletTransactionHandler federationWalletTransactionHandler,
-            IFederatedPegSettings federatedPegSettings)
+            IFederatedPegSettings federatedPegSettings,
+            IStateRepositoryRoot stateRepositoryRoot = null)
         {
             this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
             this.network = network;
             this.federationWalletManager = federationWalletManager;
             this.federationWalletTransactionHandler = federationWalletTransactionHandler;
             this.federatedPegSettings = federatedPegSettings;
+            this.stateRepositoryRoot = stateRepositoryRoot;
         }
 
         /// <inheritdoc />
-        public Transaction BuildWithdrawalTransaction(uint256 depositId, uint blockTime, Recipient recipient)
+        public BuildWithdrawalTransactionResult BuildWithdrawalTransaction(uint256 depositId, uint blockTime, Recipient recipient)
         {
             try
             {
                 this.logger.LogDebug("BuildDeterministicTransaction depositId(opReturnData)={0} recipient.ScriptPubKey={1} recipient.Amount={2}", depositId, recipient.ScriptPubKey, recipient.Amount);
 
+                // Can't send funds to known contract address.
+                if (this.stateRepositoryRoot != null)
```

Comment:
This plus the default null `StateRepositoryRoot` effectively becomes a feature toggle for skipping the recipient-is-contract test. Is that what we want?

Reply:
Yes. See above.

Reply:
Thinking some more, I'd argue that this functionality doesn't even belong in the `WithdrawalTransactionBuilder` and will cause it to violate the single responsibility principle. We should check if the address is a contract in `CrossChainTransferStore`, before attempting to build the transaction.

Reply:
Moved.